### PR TITLE
[CI, SGX] Migrate the pipeline to a patched 2.6 Intel driver

### PR DESCRIPTION
**Pull request by donporter**
## Description of the changes 

We are still using v1.9 of the Intel SGX driver; there is no good reason not to test with the latest v2.6.  

Moreover, the OOT SGX driver is not well-maintained.  To compile against a 5.8 kernel or newer, the isgx driver needs some patches.  Although I have sent a PR several weeks ago, it has gotten no attention from any maintainers.

This PR moves us to a local mirror repo, and includes the needed build fixes.  

The PR also includes an agent label for machines with 2.6 installed, so that we can gradually migrate SGX machines to v 2.6.  The host configuration and the pipeline need to move in lockstep.  Initially, we just have uncledeadly on isgx v2.6; after this lands we may need to keep one non-v2.6 worker for older PRs or ask them to rebase to master.

## How to test this PR? 

Run a jenkins pipeline.

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)

### CODE REVIEW

**Comment by donporter** [comment]
Issue with uncledeadly; stand by.

**Comment by donporter** [comment]
Retest SGX please.  Issue with uncledeadly should be fixed.

**Comment by donporter** [comment]
Jenkins, retest SGX please. Issue with uncledeadly should be fixed.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX please.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX-18.04 please.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX-18.04-apps please.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX-apps please.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX please.  Another worker setup issue.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX-apps please.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX-18.04 please.

**Comment by donporter** [comment]
Jenkins, retest Jenkins-SGX-18.04-apps please.

**Review by donporter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: 0 of 4 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @dimakuv, @mkow, and @yamahata)

___
*[Jenkinsfiles/Linux-SGX, line 29 at r1](https://reviewable.io/reviews/oscarlab/graphene/1715#-MDydUOCAVnQeXjjTuRY:-MFVycfQC7aSlmDz-7Ap:b-lah6yu) ([raw file](https://github.com/oscarlab/graphene/blob/7a7dd14b8c186a307966e2fa54a216e0f61e3159/Jenkinsfiles/Linux-SGX#L29)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I set Reviewable up, could you create a PR there?
</blockquote></details>

Done - https://github.com/oscarlab/linux-sgx-driver/pull/1

Update: Intel upstreamed the fix, so no need.  Shall I delete our mirror?

**Review by donporter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: 0 of 4 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @dimakuv, @mkow, and @yamahata)

___
*[Jenkinsfiles/Linux-SGX, line 29 at r1](https://reviewable.io/reviews/oscarlab/graphene/1715#-MDydUOCAVnQeXjjTuRY:-MGJvC_m4x_scNva16ml:b-tc2v03) ([raw file](https://github.com/oscarlab/graphene/blob/7a7dd14b8c186a307966e2fa54a216e0f61e3159/Jenkinsfiles/Linux-SGX#L29)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

Done - https://github.com/oscarlab/linux-sgx-driver/pull/1

Update: Intel upstreamed the fix, so no need.  Shall I delete our mirror?
</blockquote></details>

nm - we can keep the repo (per slack)

**Review by donporter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: 0 of 4 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @mkow)

___
*[Jenkinsfiles/Linux-SGX, line 29 at r2](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGLBYFz0fkWpTZcs07n:-MGPXc9j7EiHLkDG6vrq:brbnpn7) ([raw file](https://github.com/oscarlab/graphene/blob/a447d0cd3c24e1f948487438f345f2a8d4275d8a/Jenkinsfiles/Linux-SGX#L29)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

We shouldn't checkout master but a particular commit or tag. Otherwise our builds may suddenly break because of their fault.
</blockquote></details>

Done.  I almost did this the last time.
___
*[Jenkinsfiles/Linux-SGX-18.04, line 32 at r2](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGLBlY50IBzogCy1CCx:-MGPXgA66ncaqwuouuOh:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/a447d0cd3c24e1f948487438f345f2a8d4275d8a/Jenkinsfiles/Linux-SGX-18.04#L32)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[Jenkinsfiles/Linux-SGX-18.04-apps, line 32 at r2](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGLBmj2-2ktUHFXpIMY:-MGPXgfw6kV4F8wC7TT0:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/a447d0cd3c24e1f948487438f345f2a8d4275d8a/Jenkinsfiles/Linux-SGX-18.04-apps#L32)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[Jenkinsfiles/Linux-SGX-apps, line 29 at r2](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGLBoXt6EcNVvFhENqQ:-MGPXhHz1DxNa_-HOzhg:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/a447d0cd3c24e1f948487438f345f2a8d4275d8a/Jenkinsfiles/Linux-SGX-apps#L29)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

**Review by mkow — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: 0 of 4 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @donporter and @mkow)

___
*[Jenkinsfiles/Linux-SGX, line 29 at r2](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGLBYFz0fkWpTZcs07n:-MGPedmL7_R0GbRUZ_AQ:bgtg3vk) ([raw file](https://github.com/oscarlab/graphene/blob/a447d0cd3c24e1f948487438f345f2a8d4275d8a/Jenkinsfiles/Linux-SGX#L29)):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

Done.  I almost did this the last time.
</blockquote></details>

Please use the full hash, short hashes are trivially collidable (we even have a script for doing this automatically :) )

**Review by donporter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: 0 of 4 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @mkow)

___
*[Jenkinsfiles/Linux-SGX, line 29 at r2](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGLBYFz0fkWpTZcs07n:-MGct0XEEMdmBSeelSR6:bb6un1j) ([raw file](https://github.com/oscarlab/graphene/blob/a447d0cd3c24e1f948487438f345f2a8d4275d8a/Jenkinsfiles/Linux-SGX#L29)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please use the full hash, short hashes are trivially collidable (we even have a script for doing this automatically :) )
</blockquote></details>

Done

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @donporter)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGd6ewH4vfDorh8r3Q8:-MGd6ewH4vfDorh8r3Q9:b-f60grk) (no related file):*
Looks good modulo one thing: Won't merging it make all SGX jobs bottlenecked on the 2.6 slaves? How many of them do we have?
I don't want to suddenly make all our builds to take 2h :)

---

**Review by donporter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGd6ewH4vfDorh8r3Q8:-MGhZ2xgDwRbtVhQC-77:b-sex83p) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Looks good modulo one thing: Won't merging it make all SGX jobs bottlenecked on the 2.6 slaves? How many of them do we have?
I don't want to suddenly make all our builds to take 2h :)
</blockquote></details>

My plan was to migrate slaves onto 2.6 gradually.  Current PRs will still use the non-2.6 slaves until they rebase.

I suppose I don't have a more specific roll-out plan than monitoring latency and perhaps encouraging PRs to rebase to pick up the new 2.6 workers.  But open to suggestions.

---

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @donporter)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGd6ewH4vfDorh8r3Q8:-MGisH0v9tEREsPYZndW:bo5952d) (no related file):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

My plan was to migrate slaves onto 2.6 gradually.  Current PRs will still use the non-2.6 slaves until they rebase.

I suppose I don't have a more specific roll-out plan than monitoring latency and perhaps encouraging PRs to rebase to pick up the new 2.6 workers.  But open to suggestions.
</blockquote></details>

How many 2.6 and non-2.6 slaves do we have right now?

---

**Review by donporter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGd6ewH4vfDorh8r3Q8:-MGn98lw26Jl8Y9jGrtY:b-55anl3) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

How many 2.6 and non-2.6 slaves do we have right now?
</blockquote></details>

non-2.6: 4

2.6: just one

---

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @donporter)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1715#-MGd6ewH4vfDorh8r3Q8:-MGnBchY-tx54ow-Sy-d:belv4kv) (no related file):*
<details><summary><i>Previously, donporter (Don Porter) wrote…</i></summary><blockquote>

non-2.6: 4

2.6: just one
</blockquote></details>

Ok, so we can't merge this PR right now, it would freeze all the development.
How much time does it take to migrate one SGX worker? I'd like to have at least 3 2.6 workers before merging it.

---

**Review by mkow — APPROVED** [review]
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: :shipit: complete! all files reviewed, all discussions resolved

**Comment by mkow** [comment]
Jenkins, retest Jenkins please (it landed on a broken Jenkins worker)

**Review by dimakuv — APPROVED** [review]
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1715)* status: :shipit: complete! all files reviewed, all discussions resolved

**Comment by donporter** [comment]
Jenkins, retest Jenkins please.  beaker11 had a docker failure.

**Pull request merged**
